### PR TITLE
Stage1 mvp data

### PR DESCRIPTION
**Pull request by ryk-wolf**
------------------------------------------------------------------------------------------------------------

**Context:**
Within the Representations project, the Data classes serve as basic building blocks to allow for algebraic operations between different state representations. 

**Description of the Change:**
Implementing the Data classes which support the Representations of states that MM offers. These classes include:
- WavefunctionData
- ArrayData
- GaussianData
- SymplecticData
- QPolyData
They support a variety of algebraic operations.

**Benefits:**
Refactoring the code-base structure of MM.
Adding relevant algebras between the representations of states.

**Possible Drawbacks:**
Impossibility to use superposition of Gaussians until the batch dimension is implemented (upcoming PR).

**Related GitHub Issues:**
N.A.

### CODE REVIEW

File: mrmustard/lab/representations/data/gaussian_data.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+    @property
+    def means(self) -> Vector:
+        return self.vec
+
+    @property
+    def c(self) -> Scalar:
+        return self.coeffs
+
+    def __mul__(self, other: Union[Scalar, GaussianData]) -> GaussianData:
```
Comment (sylviemonet) [inline mrmustard/lab/representations/data/gaussian_data.py:98]: Wait for the new formula for this

File: mrmustard/lab/representations/data/matvec_data.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        try:
+            if super().same(X=[self.mat, self.vec], Y=[other.mat, other.vec]):
+                combined_coeffs = self.coeffs + other.coeffs
+                return self.__class__(self.mat, self.vec, combined_coeffs)
+
+            else:
+                mat = math.concat([self.mat, other.mat], axis=0)
+                vec = math.concat([self.vec, other.vec], axis=0)
+                reorder_matrix = reorder_matrix_from_qpqp_to_qqpp(self.mat.shape[-1])
+                mat = math.matmul(math.matmul(reorder_matrix, mat), math.transpose(reorder_matrix))
```
Comment (sylviemonet) [inline mrmustard/lab/representations/data/matvec_data.py:64]: I have no idea of this matmul...

File: mrmustard/lab/representations/data/matvec_data.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+                vec = math.matvec(reorder_matrix, vec)
```
Comment (sylviemonet) [inline mrmustard/lab/representations/data/matvec_data.py:65]: I have no idea of this matvec either...

File: mrmustard/lab/representations/data/symplectic_data.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        self, symplectic: RealMatrix, displacement: RealVector, coeffs: Scalar = 1.0
+    ) -> None:
+        # Check if it is a symplectic matrix
+        if is_symplectic(math.asnumpy(symplectic)):
+            super().__init__(mat=symplectic, vec=displacement, coeffs=coeffs)
+        else:
+            raise ValueError("The matrix given is not symplectic.")
+
+    @property
+    def symplectic(self) -> np.array:
```
Comment (sylviemonet) [inline mrmustard/lab/representations/data/symplectic_data.py:49]: np.array would be RealMatrix?

File: mrmustard/lab/representations/data/symplectic_data.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        return self.mat
+
+    @property
+    def displacement(self) -> np.array:
```
Comment (sylviemonet) [inline mrmustard/lab/representations/data/symplectic_data.py:53]: np.array would be RealVector?

File: mrmustard/lab/representations/data/symplectic_data.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        return self.vec
+
+    def __mul__(self, other: Scalar) -> SymplecticData:
+        if isinstance(other, SymplecticData):
+            raise TypeError("Symplectic can only be multiplied by a scalar")
```
Comment (sylviemonet) [inline mrmustard/lab/representations/data/symplectic_data.py:58]: "We do not support multiplication of two symplectic data now."

**Review by sylviemonet — CHANGES_REQUESTED** [review]

File: mrmustard/lab/representations/data/gaussian_data.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+import numpy as np
+
+from typing import Optional, TYPE_CHECKING, Union
+
+from mrmustard.lab.representations.data.matvec_data import MatVecData
+from mrmustard.math import Math
+from mrmustard.typing import Matrix, Scalar, Tensor, Vector
+
+
+if TYPE_CHECKING:  # This is to avoid the circular import issu with GaussianData<>QPolyData
```
Comment (sylviemonet) [inline mrmustard/lab/representations/data/gaussian_data.py:26]: This type checking is not needed anymore.

File: mrmustard/lab/representations/data/matvec_data.py (hunk 2 above)
Comment (ryk-wolf) [reply mrmustard/lab/representations/data/matvec_data.py:64]: OK, so what should we do about this? Shall we ask Seb/Filippo to review it too?

File: mrmustard/lab/representations/data/matvec_data.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    def __eq__(self, other: MatVecData) -> bool:
+        try:
+            return super().same(
+                X=[self.mat, self.vec, self.coeffs], Y=[other.mat, other.vec, other.coeffs]
+            )
+
+        except AttributeError as e:
+            raise TypeError(f"Cannot compare {self.__class__} and {other.__class__}.") from e
+
+    def __add__(self, other: MatVecData) -> MatVecData:
```
Comment (ryk-wolf) [inline mrmustard/lab/representations/data/matvec_data.py:54]: Get rid of function

File: mrmustard/lab/representations/data/matvec_data.py (hunk 2 above)
Comment (ryk-wolf) [reply mrmustard/lab/representations/data/matvec_data.py]: Sorted

File: mrmustard/lab/representations/data/matvec_data.py (hunk 3 above)
Comment (ryk-wolf) [reply mrmustard/lab/representations/data/matvec_data.py]: Sorted

File: mrmustard/lab/representations/data/gaussian_data.py (hunk 7 above)
Comment (ryk-wolf) [reply mrmustard/lab/representations/data/gaussian_data.py]: done

File: mrmustard/lab/representations/data/symplectic_data.py (hunk 5 above)
Comment (ryk-wolf) [reply mrmustard/lab/representations/data/symplectic_data.py]: done

File: mrmustard/lab/representations/data/symplectic_data.py (hunk 4 above)
Comment (ryk-wolf) [reply mrmustard/lab/representations/data/symplectic_data.py]: done

File: mrmustard/lab/representations/data/symplectic_data.py (hunk 6 above)
Comment (sylviemonet) [reply mrmustard/lab/representations/data/symplectic_data.py:58]: For symplectic data, no

Comment (ryk-wolf) [reply mrmustard/lab/representations/data/symplectic_data.py]: Ok. So I'm not sure why the current error message is wrong. ^^'

Comment (sylviemonet) [reply mrmustard/lab/representations/data/symplectic_data.py]: Ah it is not wrong. Just provide another way to say the sentence. That's why I use "" to quote.

**Comment by sylviemonet** [comment]
Test has not passed now. Can you have a look @ryk-wolf please?

**Comment by ryk-wolf** [comment]
> 

Done, tests pass now.

File: mrmustard/lab/representations/data/data.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    @abstractmethod
+    def __truediv__(self, other: Union[Scalar, Data]) -> Data:
+        raise NotImplementedError()  # prompting override in children
+
+    @abstractmethod
+    def __mul__(self, other: Union[Scalar, Data]) -> Data:
+        raise NotImplementedError()  # prompting override in children
+
+    def __rmul__(self, other: Scalar) -> Data:
+        return self.__mul__(other=other)
```
Comment (ziofil) [inline mrmustard/lab/representations/data/data.py:75]: is this this or `other.__mul__(self)`?

File: mrmustard/lab/representations/data/data.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+    # @abstractmethod
```
Comment (ziofil) [inline mrmustard/lab/representations/data/data.py:77]: if it needs to go delete, if it should stay leave it with `NotImplementedError` at least

**Pull request closed without merging**